### PR TITLE
add bert model

### PR DESCRIPTION
**Pull request by L1aoXingyu**
利用 layers 中的积木构建一个 bert 模型，支持 2d sbp 混合并行

### CODE REVIEW

File: libai/layers/cross_entropy.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import oneflow as flow
+from oneflow import nn
+from libai.utils import distributed as dist
+
+
+class ParallelCrossEntropyLossWithMask(nn.Module):
```
Comment (dangkai4u) [inline libai/layers/cross_entropy.py:22]: 损失函数尽量和pytorch接口类似，只输入logits和labels，不输入loss_mask，通过输入ignore_index避免对pad以及未被mask单词计算损失。

前几天仿写了一个文件，可供参考：https://github.com/Oneflow-Inc/libai/blob/tutorial/tutorial.py#L345 。

不过这种方法是在运行中计算loss_mask，和数据处理阶段生成loss_mask不同，不确定哪种开销更大。但至少这个类里不要loss_mask，可以放在bert_loss那里，让这个类尽可能成为通用模块。

ps: oneflow里的交叉熵损失有好几种写法，比较冗余，希望能反馈给框架组，精简一下

File: libai/models/bert_model.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+    ParallelCrossEntropyLossWithMask,
+)
+from libai.utils import distributed as dist
+from oneflow import nn
+from libai.config import configurable
+
+from .utils import init_method_normal, scaled_init_method_normal
+
+
+class BertExtendedAttnMask(nn.Module):
```
Comment (dangkai4u) [inline libai/models/bert_model.py:34]: 之前讨论过一次attention_mask所有操作抽象为一个模块，当时没有讨论出结果，最近发现，transformers里相当于统一了。虽然它的实现方式不值得我们学习，但至少证明了，attention_mask操作就那么几类，且不同的模型可以共用，我建议大家讨论一下这个部分，抽象成一个类，作为基础模块，供所有模型使用。

transformer: https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_utils.py#L244

之前我归纳的：https://github.com/Oneflow-Inc/libai/blob/dist_gpt/libai/layers/mask_helper.py

File: libai/models/bert_model.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+        if self.tokentype_embeddings is not None:
+            if tokentype_ids is None:
+                tokentype_ids = (
+                    self.tokentype_ids[:, :seq_length]
+                    .expand_as(input_ids)
+                    .to_consistent(sbp=input_ids.sbp)
+                )
+            embeddings = embeddings + self.tokentype_embeddings(tokentype_ids)
+
+        embeddings = self.embedding_dropout(embeddings)
```
Comment (dangkai4u) [inline libai/models/bert_model.py:120]: bert的embedding，先经过layernorm，再经过dropout。

File: libai/models/bert_model.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+        return embeddings
+
+    def word_embeddings(self):
+        return self.vocab_embeddings.weight
+
+
+class BertLMPredictionHead(nn.Module):
+    def __init__(self, hidden_size, init_method, bias_gelu_fusion=True):
```
Comment (dangkai4u) [inline libai/models/bert_model.py:128]: 后面这些head，并行方式测试过了吗，不确定应该采用哪种，看megatron代码，到了这里就不使用模型并行了。这里不要加fusion了，因为这部分代码不会用于预训练，效率不是主要问题，尽量让这里简单，使用体验和普通的linear一样。

File: libai/models/bert_model.py [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+            hidden_size,
+            bias=True,
+            parallel="col",
+            init_method=init_method,
+            layer_idx=-1,
+        )
+        self.activation_func = build_activation("tanh")
+
+    def forward(self, hidden_states, sequence_index=0):
+        """Just "pool" the model by simply taking the [CLS] token corresponding to the first token.
```
Comment (dangkai4u) [inline libai/models/bert_model.py:187]: 看其他代码，都是pooling第一个元素，没有选择一个位置pooling的情况。其他的pooling方式就是mean或max，所以这里就不用加sequence_index参数了

File: libai/models/bert_model.py [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+            scale_mask_softmax_fusion,
+            apply_query_key_layer_scaling,
+        )
+
+        self.pooler = (
+            BertPooler(hidden_size, init_method) if add_pooling_layer else None
+        )
+
+    @classmethod
+    def from_config(cls, cfg):
```
Comment (dangkai4u) [inline libai/models/bert_model.py:415]: 这个函数返回值为什么不是cls，而是dict呢，它的意思应该是输入cfg，创建个类啊

File: libai/models/bert_model.py [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+            layer_idx=-1,
+        )
+
+    def forward(self, sequence_output, pooled_output):
+        prediction_scores = self.predictions(sequence_output)
+        seq_relationship_score = self.seq_relationship(pooled_output)
+        return prediction_scores, seq_relationship_score
+
+
+class LMLogits(nn.Module):
```
Comment (dangkai4u) [inline libai/models/bert_model.py:217]: 这个类放到layers吧，应该也算是通用的模块，这种tie_embedding的方式，没有bias参数。另外，我在bert里没有找到使用它的情形，gpt里倒是用过这个。

File: libai/models/bert_model.py [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+            vocab_size, hidden_size, init_method=init_method
+        )
+        self.position_embeddings = Embedding(
+            max_sequence_length, hidden_size, init_method=init_method
+        )
+
+        # NOTE(l1aoxingyu): Set position_ids sbp sign to [B, B] initially, because position_ids is a
+        # 1D-tensor from 0 to seq_length, if set to [S(0), B] at first, then position_ids
+        # will split at the first dim of hierarchy.
+        self.position_ids = flow.arange(
```
Comment (dangkai4u) [inline libai/models/bert_model.py:75]: 对1D-tensor传入[B, B]也可以吗

File: libai/models/bert_model.py (hunk 2 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:34]: 这个问题提一个新的 pr 来推进吧

File: libai/models/bert_model.py (hunk 3 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:120]: 这里是对齐的 megatron 的实现，你再 check 一下呢
https://github.com/NVIDIA/Megatron-LM/blob/9d86ca677c48b4e006156b1edce631e6c4bd5c12/megatron/model/language_model.py#L193-L205

File: libai/models/bert_model.py (hunk 4 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:128]: 这个部分的代码不会用于预训练是什么意思呀？是指不参与训练吗？

File: libai/models/bert_model.py (hunk 6 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:415]: 因为 `__init__` 上面有一个 decorator `configurable`，这时这个类可以通过下面两种方式进行构造

`BertModel(cfg)` 或者是 `BertModel(100, 756, ...)` 或者是混合传参 `BertModel(cfg, layernorm_eps=1e-4)`

而 `BertModel(cfg)` 实际调用的方法就是 `from_config`

File: libai/models/bert_model.py (hunk 7 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:217]: 这个是在 bert pretraining 里面算 mask loss 之前用的

File: libai/models/bert_model.py (hunk 8 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:75]: 可以传 b, b

File: libai/models/bert_model.py (hunk 4 above)
Comment (dangkai4u) [reply libai/models/bert_model.py:128]: 我看错了。我参考的原始bert，你做对齐就按照这样吧。不过megatron里没有使用fusion，fusion尽量只在transformer_layer里使用，外面用普通的linear+gelu吧。

File: libai/models/bert_model.py (hunk 3 above)
Comment (dangkai4u) [reply libai/models/bert_model.py:120]: 对齐的话，那就按这样来吧，不同的库实现bert的细节不一样，等对齐之后这些再统一

File: libai/models/bert_model.py [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+            max_sequence_length,
+            dtype=flow.long,
+            sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
+            placement=dist.get_layer_placement(0),
+        ).unsqueeze(0)
```
Comment (CPFLAME) [inline libai/models/bert_model.py:80]: 我记得之前oneflow在生成consistent的range时用的是`flow._C.consistent_arange`,  现在可以直接用`flow.arange`了?

File: libai/layers/cross_entropy.py (hunk 1 above)
Comment (L1aoXingyu) [reply libai/layers/cross_entropy.py:22]: 现在用的 megatron 的 dataloader 是在数据读取阶段就生成了 loss_mask，那就把 loss_mask 放到 bert_loss 那里去计算，这里直接 return sparse_softmax_cross_entropy 的结果

File: libai/models/bert_model.py (hunk 7 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:217]: 这里 bias 作为一个 optional 参数

https://github.com/NVIDIA/Megatron-LM/blob/9d86ca677c48b4e006156b1edce631e6c4bd5c12/megatron/model/language_model.py#L29-L39

File: libai/models/bert_model.py (hunk 9 above)
Comment (L1aoXingyu) [reply libai/models/bert_model.py:80]: 对，现在可以直接在 flow.arange 里面指定 sbp 和 placement

File: libai/models/bert_model.py [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+            init_method=init_method,
+            layer_idx=-1,
+        )
+        self.activation_func = build_activation("tanh")
+
+    def forward(self, hidden_states):
+        """Just "pool" the model by simply taking the [CLS] token corresponding to the first token.
+        """
+        # hidden_states: [bsz, seq_len, hidden_size]
+        select_token_tensor = hidden_states[:, 0]
```
Comment (CPFLAME) [inline libai/models/bert_model.py:181]: 这里要不要换成 `hidden_states[:, 0, :]`?

Comment (L1aoXingyu) [reply libai/models/bert_model.py]: 改了

**Review by CPFLAME — APPROVED** [review]
